File: src/lib/tauri/interceptor.ts
// SPDX-License-Identifier: Apache-2.0

// TypeScript bindings for the Rust Universal Query Interceptor. All data is
// stored and processed in the backend.

import { invoke } from '@/lib/transport';

export type Environment = 'development' | 'staging' | 'production';

export type QueryOperationType =
  | 'select'
  | 'insert'
  | 'update'
  | 'delete'
  | 'create'
  | 'alter'
  | 'drop'
  | 'truncate'
  | 'grant'
  | 'revoke'
  | 'execute'
  | 'other';

export type SafetyAction = 'block' | 'warn' | 'require_confirmation';

export interface SafetyRule {
  id: string;
  name: string;
  description: string;
  enabled: boolean;
  environments: Environment[];
  operations: QueryOperationType[];
  action: SafetyAction;
  pattern?: string;
  builtin: boolean;
}

export const BUILTIN_SAFETY_RULE_I18N: Record<string, { nameKey: string; descriptionKey: string }> =
  {
    'builtin-no-drop-production': {
      nameKey: 'interceptor.safety.builtinRuleNames.builtin-no-drop-production',
      descriptionKey: 'interceptor.safety.builtinRuleDescriptions.builtin-no-drop-production',
    },
    'builtin-no-truncate-production': {
      nameKey: 'interceptor.safety.builtinRuleNames.builtin-no-truncate-production',
      descriptionKey: 'interceptor.safety.builtinRuleDescriptions.builtin-no-truncate-production',
    },
    'builtin-confirm-delete-production': {
      nameKey: 'interceptor.safety.builtinRuleNames.builtin-confirm-delete-production',
      descriptionKey:
        'interceptor.safety.builtinRuleDescriptions.builtin-confirm-delete-production',
    },
    'builtin-confirm-update-no-where': {
      nameKey: 'interceptor.safety.builtinRuleNames.builtin-confirm-update-no-where',
      descriptionKey: 'interceptor.safety.builtinRuleDescriptions.builtin-confirm-update-no-where',
    },
    'builtin-confirm-delete-no-where': {
      nameKey: 'interceptor.safety.builtinRuleNames.builtin-confirm-delete-no-where',
      descriptionKey: 'interceptor.safety.builtinRuleDescriptions.builtin-confirm-delete-no-where',
    },
    'builtin-warn-alter-production': {
      nameKey: 'interceptor.safety.builtinRuleNames.builtin-warn-alter-production',
      descriptionKey: 'interceptor.safety.builtinRuleDescriptions.builtin-warn-alter-production',
    },
    n_plus_one: {
      nameKey: 'interceptor.safety.builtinRuleNames.n_plus_one',
      descriptionKey: 'interceptor.safety.builtinRuleDescriptions.n_plus_one',
    },
    alert_error_rate: {
      nameKey: 'interceptor.safety.builtinRuleNames.alert_error_rate',
      descriptionKey: 'interceptor.safety.builtinRuleDescriptions.alert_error_rate',
    },
    alert_slow_queries: {
      nameKey: 'interceptor.safety.builtinRuleNames.alert_slow_queries',
      descriptionKey: 'interceptor.safety.builtinRuleDescriptions.alert_slow_queries',
    },
  };

export interface AuditLogEntry {
  id: string;
  timestamp: string;
  session_id: string;
  query: string;
  query_preview: string;
  environment: Environment;
  operation_type: QueryOperationType;
  database?: string;
  success: boolean;
  error?: string;
  execution_time_ms: number;
  row_count?: number;
  blocked: boolean;
  safety_rule?: string;
  driver_id: string;
  fingerprint?: string;
}

export type AuditExportFormat = 'json' | 'jsonl' | 'csv';

export interface AuditStats {
  total: number;
  successful: number;
  failed: number;
  blocked: number;
  last_hour: number;
  last_day: number;
  by_environment: Record<string, number>;
  by_operation: Record<string, number>;
}

export interface ProfilingMetrics {
  total_queries: number;
  successful_queries: number;
  failed_queries: number;
  blocked_queries: number;
  total_execution_time_ms: number;
  avg_execution_time_ms: number;
  min_execution_time_ms: number;
  max_execution_time_ms: number;
  p50_execution_time_ms: number;
  p95_execution_time_ms: number;
  p99_execution_time_ms: number;
  slow_query_count: number;
  by_operation_type: Record<string, number>;
  by_environment: Record<string, number>;
  period_start: string;
}

export interface SlowQueryEntry {
  id: string;
  timestamp: string;
  query: string;
  execution_time_ms: number;
  environment: Environment;
  database?: string;
  row_count?: number;
  driver_id: string;
}

export interface InterceptorConfig {
  audit_enabled: boolean;
  profiling_enabled: boolean;
  safety_enabled: boolean;
  slow_query_threshold_ms: number;
  max_audit_entries: number;
  max_slow_queries: number;
  safety_rules: SafetyRule[];
  builtin_rule_overrides: BuiltinRuleOverride[];
  /** Error-rate alert over 15 minutes, in percent. Absent or 0 disables it. */
  alert_error_rate_percent?: number | null;
  /** Slow-query count alert over 15 minutes. Absent or 0 disables it. */
  alert_slow_queries_count?: number | null;
}

export interface TrendPoint {
  day: string;
  count: number;
  p50_ms: number;
  p95_ms: number;
  error_rate: number;
}

export interface Regression {
  recent_p95_ms: number;
  baseline_p95_ms: number;
  recent_count: number;
}

export interface FingerprintTrend {
  fingerprint: string;
  query_preview: string;
  driver_id: string;
  database?: string | null;
  count: number;
  p50_ms: number;
  p95_ms: number;
  error_rate: number;
  points: TrendPoint[];
  regression?: Regression;
}

export interface TrendFilter {
  days: number;
  driver_id?: string;
  database?: string;
}

export type InterceptorAlert =
  | {
      kind: 'n_plus_one';
      session_id: string;
      fingerprint: string;
      query_preview: string;
      count: number;
    }
  | { kind: 'error_rate'; percent: number; threshold: number; total: number }
  | { kind: 'slow_queries'; count: number; threshold: number };

export interface BuiltinRuleOverride {
  id: string;
  enabled: boolean;
}

export interface AuditFilter {
  limit?: number;
  offset?: number;
  environment?: Environment;
  operation?: QueryOperationType;
  success?: boolean;
  search?: string;
  fingerprint?: string;
  blocked?: boolean;
}

interface InterceptorConfigResponse {
  success: boolean;
  config?: InterceptorConfig;
  error?: string;
}

interface AuditEntriesResponse {
  success: boolean;
  entries: AuditLogEntry[];
  error?: string;
}

interface AuditStatsResponse {
  success: boolean;
  stats?: AuditStats;
  error?: string;
}

interface ProfilingMetricsResponse {
  success: boolean;
  metrics?: ProfilingMetrics;
  error?: string;
}

interface SlowQueriesResponse {
  success: boolean;
  queries: SlowQueryEntry[];
  error?: string;
}

interface SafetyRulesResponse {
  success: boolean;
  rules: SafetyRule[];
  error?: string;
}

interface GenericResponse {
  success: boolean;
  error?: string;
}

interface ExportResponse {
  success: boolean;
  data?: string;
  error?: string;
}

export async function getInterceptorConfig(): Promise<InterceptorConfig> {
  const result = await invoke<InterceptorConfigResponse>('get_interceptor_config');
  if (!result.success || !result.config) {
    throw new Error(result.error || 'Failed to get interceptor config');
  }
  return result.config;
}

export async function updateInterceptorConfig(
  config: InterceptorConfig
): Promise<InterceptorConfig> {
  const result = await invoke<InterceptorConfigResponse>('update_interceptor_config', { config });
  if (!result.success || !result.config) {
    throw new Error(result.error || 'Failed to update interceptor config');
  }
  return result.config;
}

export async function getAuditEntries(filter: AuditFilter = {}): Promise<AuditLogEntry[]> {
  const result = await invoke<AuditEntriesResponse>('get_audit_entries', { filter });
  if (!result.success) {
    throw new Error(result.error || 'Failed to get audit entries');
  }
  return result.entries;
}

export async function getAuditStats(): Promise<AuditStats> {
  const result = await invoke<AuditStatsResponse>('get_audit_stats');
  if (!result.success || !result.stats) {
    throw new Error(result.error || 'Failed to get audit stats');
  }
  return result.stats;
}

/**
 * Clear the audit log. Acquires a one-shot confirmation token first so a
 * drive-by IPC call cannot wipe the audit trail.
 */
export async function clearAuditLog(): Promise<void> {
  const { token } = await invoke<{ token: string; expires_in_secs: number }>(
    'request_confirmation_token',
    { action: 'clear_audit_log' }
  );
  const result = await invoke<GenericResponse>('clear_audit_log', {
    confirmationToken: token,
  });
  if (!result.success) {
    throw new Error(result.error || 'Failed to clear audit log');
  }
}

/**
 * Export audit log in the requested format.
 *
 * `fromDisk = true` reads the full retained history from the rotated JSONL
 * file rather than the in-memory cache — needed when the user wants a
 * faithful audit trail beyond the current cache window.
 */
export async function exportAuditLog(
  format: AuditExportFormat = 'json',
  fromDisk = false
): Promise<string> {
  const result = await invoke<ExportResponse>('export_audit_log', {
    format,
    fromDisk,
  });
  if (!result.success || !result.data) {
    throw new Error(result.error || 'Failed to export audit log');
  }
  return result.data;
}

export async function getProfilingMetrics(): Promise<ProfilingMetrics> {
  const result = await invoke<ProfilingMetricsResponse>('get_profiling_metrics');
  if (!result.success || !result.metrics) {
    throw new Error(result.error || 'Failed to get profiling metrics');
  }
  return result.metrics;
}

export async function getQueryTrends(filter: TrendFilter): Promise<FingerprintTrend[]> {
  const response = await invoke<{
    success: boolean;
    trends: FingerprintTrend[];
    error?: string;
  }>('get_query_trends', { filter });
  if (!response.success) {
    throw new Error(response.error || 'Failed to load query trends');
  }
  return response.trends;
}

export async function getSlowQueries(limit = 50, offset = 0): Promise<SlowQueryEntry[]> {
  const result = await invoke<SlowQueriesResponse>('get_slow_queries', { limit, offset });
  if (!result.success) {
    throw new Error(result.error || 'Failed to get slow queries');
  }
  return result.queries;
}

export async function clearSlowQueries(): Promise<void> {
  const result = await invoke<GenericResponse>('clear_slow_queries');
  if (!result.success) {
    throw new Error(result.error || 'Failed to clear slow queries');
  }
}

export async function resetProfilingMetrics(): Promise<void> {
  const result = await invoke<GenericResponse>('reset_profiling');
  if (!result.success) {
    throw new Error(result.error || 'Failed to reset profiling');
  }
}

export async function exportProfilingData(): Promise<string> {
  const result = await invoke<ExportResponse>('export_profiling');
  if (!result.success || !result.data) {
    throw new Error(result.error || 'Failed to export profiling data');
  }
  return result.data;
}

export async function getSafetyRules(): Promise<SafetyRule[]> {
  const result = await invoke<SafetyRulesResponse>('get_safety_rules');
  if (!result.success) {
    throw new Error(result.error || 'Failed to get safety rules');
  }
  return result.rules;
}

export async function addSafetyRule(rule: SafetyRule): Promise<SafetyRule[]> {
  const result = await invoke<SafetyRulesResponse>('add_safety_rule', { rule });
  if (!result.success) {
    throw new Error(result.error || 'Failed to add safety rule');
  }
  return result.rules;
}

export async function updateSafetyRule(rule: SafetyRule): Promise<SafetyRule[]> {
  const result = await invoke<SafetyRulesResponse>('update_safety_rule', { rule });
  if (!result.success) {
    throw new Error(result.error || 'Failed to update safety rule');
  }
  return result.rules;
}

export async function removeSafetyRule(ruleId: string): Promise<SafetyRule[]> {
  const result = await invoke<SafetyRulesResponse>('remove_safety_rule', { ruleId });
  if (!result.success) {
    throw new Error(result.error || 'Failed to remove safety rule');
  }
  return result.rules;
}

export function formatExecutionTime(ms: number): string {
  if (ms < 1) {
    return `${(ms * 1000).toFixed(0)}µs`;
  } else if (ms < 1000) {
    return `${ms.toFixed(1)}ms`;
  } else {
    return `${(ms / 1000).toFixed(2)}s`;
  }
}

export function getPerformanceClass(ms: number): 'fast' | 'normal' | 'slow' | 'critical' {
  if (ms < 100) return 'fast';
  if (ms < 500) return 'normal';
  if (ms < 2000) return 'slow';
  return 'critical';
}

export function getPerformanceColor(perfClass: 'fast' | 'normal' | 'slow' | 'critical'): string {
  switch (perfClass) {
    case 'fast':
      return '#22c55e'; // green-500
    case 'normal':
      return '#3b82f6'; // blue-500
    case 'slow':
      return '#f59e0b'; // amber-500
    case 'critical':
      return '#ef4444'; // red-500
  }
}

export interface GovernanceLimits {
  max_query_duration_ms: number | null;
  max_result_rows: number | null;
  max_concurrent_queries: number | null;
}

export async function getGovernanceLimits(): Promise<GovernanceLimits> {
  return await invoke<GovernanceLimits>('get_governance_limits');
}

export async function updateGovernanceLimits(limits: GovernanceLimits): Promise<GovernanceLimits> {
  return await invoke<GovernanceLimits>('update_governance_limits', { limits });
}

File: src-tauri/src/commands/interceptor.rs
// SPDX-License-Identifier: Apache-2.0

//! Commands for managing the Universal Query Interceptor system.

use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tauri::State;

use crate::interceptor::{
    AuditExportFormat, AuditLogEntry, AuditStats, Environment, FingerprintTrend, InterceptorConfig,
    ProfilingMetrics, QueryOperationType, SafetyRule, SlowQueryEntry, TrendFilter,
};

#[derive(Debug, Serialize)]
pub struct QueryTrendsResponse {
    pub success: bool,
    pub trends: Vec<FingerprintTrend>,
    pub error: Option<String>,
}

/// Core: per-fingerprint trends from the audit file. Regressions ride along
/// only in Pro builds.
#[tauri::command]
pub async fn get_query_trends(
    state: State<'_, crate::SharedState>,
    filter: TrendFilter,
) -> Result<QueryTrendsResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };
    Ok(
        match interceptor.get_query_trends(&filter, cfg!(feature = "pro")) {
            Ok(trends) => QueryTrendsResponse {
                success: true,
                trends,
                error: None,
            },
            Err(error) => QueryTrendsResponse {
                success: false,
                trends: Vec::new(),
                error: Some(error),
            },
        },
    )
}

#[derive(Debug, Serialize)]
pub struct InterceptorConfigResponse {
    pub success: bool,
    pub config: Option<InterceptorConfig>,
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct AuditEntriesResponse {
    pub success: bool,
    pub entries: Vec<AuditLogEntry>,
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct AuditStatsResponse {
    pub success: bool,
    pub stats: Option<AuditStats>,
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ProfilingMetricsResponse {
    pub success: bool,
    pub metrics: Option<ProfilingMetrics>,
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct SlowQueriesResponse {
    pub success: bool,
    pub queries: Vec<SlowQueryEntry>,
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct SafetyRulesResponse {
    pub success: bool,
    pub rules: Vec<SafetyRule>,
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct GenericResponse {
    pub success: bool,
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ExportResponse {
    pub success: bool,
    pub data: Option<String>,
    pub error: Option<String>,
}

#[tauri::command]
pub async fn get_interceptor_config(
    state: State<'_, crate::SharedState>,
) -> Result<InterceptorConfigResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };

    let config = interceptor.get_config();

    Ok(InterceptorConfigResponse {
        success: true,
        config: Some(config),
        error: None,
    })
}

#[tauri::command]
pub async fn update_interceptor_config(
    state: State<'_, crate::SharedState>,
    config: InterceptorConfig,
) -> Result<InterceptorConfigResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };

    match interceptor.update_config(config) {
        Ok(()) => {
            let updated = interceptor.get_config();
            Ok(InterceptorConfigResponse {
                success: true,
                config: Some(updated),
                error: None,
            })
        }
        Err(e) => Ok(InterceptorConfigResponse {
            success: false,
            config: None,
            error: Some(e),
        }),
    }
}

#[derive(Debug, Deserialize)]
pub struct AuditFilter {
    pub limit: Option<usize>,
    pub offset: Option<usize>,
    pub environment: Option<Environment>,
    pub operation: Option<QueryOperationType>,
    pub success: Option<bool>,
    pub search: Option<String>,
    /// Restrict results to entries matching this fingerprint (Pro).
    pub fingerprint: Option<String>,
    /// `Some(true)` keeps only blocked entries; `Some(false)` excludes them.
    pub blocked: Option<bool>,
}

/// Core: limited to 50 entries, no advanced filters. Pro: unlimited.
#[tauri::command]
pub async fn get_audit_entries(
    state: State<'_, crate::SharedState>,
    filter: AuditFilter,
) -> Result<AuditEntriesResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };

    #[cfg(feature = "pro")]
    let entries = interceptor.get_audit_entries(
        filter.limit.unwrap_or(100),
        filter.offset.unwrap_or(0),
        filter.environment,
        filter.operation,
        filter.success,
        filter.search.as_deref(),
        filter.fingerprint.as_deref(),
        filter.blocked,
    );

    #[cfg(not(feature = "pro"))]
    let entries = interceptor.get_audit_entries(
        filter.limit.unwrap_or(50).min(50),
        filter.offset.unwrap_or(0),
        None, // No environment filter in Core
        None, // No operation filter in Core
        None, // No success filter in Core
        None, // No search in Core
        None, // No fingerprint filter in Core
        None, // No blocked filter in Core
    );

    Ok(AuditEntriesResponse {
        success: true,
        entries,
        error: None,
    })
}

#[tauri::command]
pub async fn get_audit_stats(
    state: State<'_, crate::SharedState>,
) -> Result<AuditStatsResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };

    let stats = interceptor.get_audit_stats();

    Ok(AuditStatsResponse {
        success: true,
        stats: Some(stats),
        error: None,
    })
}

/// Clears the audit log. Requires a one-shot confirmation token issued by
/// `request_confirmation_token("clear_audit_log")` to prevent drive-by IPC
/// calls from destroying the audit trail (SOC2 / RGPD impact).
#[tauri::command]
pub async fn clear_audit_log(
    state: State<'_, crate::SharedState>,
    confirmation_token: String,
) -> Result<GenericResponse, String> {
    let (interceptor, confirmation_tokens) = {
        let state = state.lock().await;
        (
            Arc::clone(&state.interceptor),
            Arc::clone(&state.confirmation_tokens),
        )
    };

    confirmation_tokens.consume("clear_audit_log", &confirmation_token)?;
    interceptor.clear_audit();
    tracing::warn!("audit log cleared via clear_audit_log");

    Ok(GenericResponse {
        success: true,
        error: None,
    })
}

/// Exports the audit log (Pro only).
///
/// `format` selects the serialization (`json`, `jsonl`, `csv`). When
/// `from_disk` is `true`, the entire retained history is read from the rotated
/// JSONL file rather than just the in-memory cache — useful when retention
/// exceeds the cache size.
#[cfg(feature = "pro")]
#[tauri::command]
pub async fn export_audit_log(
    state: State<'_, crate::SharedState>,
    format: Option<AuditExportFormat>,
    from_disk: Option<bool>,
) -> Result<ExportResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };

    let format = format.unwrap_or_default();
    let from_disk = from_disk.unwrap_or(false);

    match interceptor.export_audit_format(format, from_disk) {
        Ok(data) => Ok(ExportResponse {
            success: true,
            data: Some(data),
            error: None,
        }),
        Err(e) => Ok(ExportResponse {
            success: false,
            data: None,
            error: Some(e),
        }),
    }
}

#[cfg(not(feature = "pro"))]
#[tauri::command]
pub async fn export_audit_log(
    _state: State<'_, crate::SharedState>,
    _format: Option<AuditExportFormat>,
    _from_disk: Option<bool>,
) -> Result<ExportResponse, String> {
    Ok(ExportResponse {
        success: false,
        data: None,
        error: Some("Audit log export requires QoreDB Pro".into()),
    })
}

#[cfg(feature = "pro")]
#[tauri::command]
pub async fn get_profiling_metrics(
    state: State<'_, crate::SharedState>,
) -> Result<ProfilingMetricsResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };
    let metrics = interceptor.get_profiling_metrics();
    Ok(ProfilingMetricsResponse {
        success: true,
        metrics: Some(metrics),
        error: None,
    })
}

#[cfg(not(feature = "pro"))]
#[tauri::command]
pub async fn get_profiling_metrics(
    _state: State<'_, crate::SharedState>,
) -> Result<ProfilingMetricsResponse, String> {
    Ok(ProfilingMetricsResponse {
        success: false,
        metrics: None,
        error: Some("Query profiling requires QoreDB Pro".into()),
    })
}

#[cfg(feature = "pro")]
#[tauri::command]
pub async fn get_slow_queries(
    state: State<'_, crate::SharedState>,
    limit: Option<usize>,
    offset: Option<usize>,
) -> Result<SlowQueriesResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };
    let queries = interceptor.get_slow_queries(limit.unwrap_or(50), offset.unwrap_or(0));
    Ok(SlowQueriesResponse {
        success: true,
        queries,
        error: None,
    })
}

#[cfg(not(feature = "pro"))]
#[tauri::command]
pub async fn get_slow_queries(
    _state: State<'_, crate::SharedState>,
    _limit: Option<usize>,
    _offset: Option<usize>,
) -> Result<SlowQueriesResponse, String> {
    Ok(SlowQueriesResponse {
        success: false,
        queries: vec![],
        error: Some("Query profiling requires QoreDB Pro".into()),
    })
}

#[cfg(feature = "pro")]
#[tauri::command]
pub async fn clear_slow_queries(
    state: State<'_, crate::SharedState>,
) -> Result<GenericResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };
    interceptor.clear_slow_queries();
    Ok(GenericResponse {
        success: true,
        error: None,
    })
}

#[cfg(not(feature = "pro"))]
#[tauri::command]
pub async fn clear_slow_queries(
    _state: State<'_, crate::SharedState>,
) -> Result<GenericResponse, String> {
    Ok(GenericResponse {
        success: false,
        error: Some("Query profiling requires QoreDB Pro".into()),
    })
}

#[cfg(feature = "pro")]
#[tauri::command]
pub async fn reset_profiling(
    state: State<'_, crate::SharedState>,
) -> Result<GenericResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };
    interceptor.reset_profiling();
    Ok(GenericResponse {
        success: true,
        error: None,
    })
}

#[cfg(not(feature = "pro"))]
#[tauri::command]
pub async fn reset_profiling(
    _state: State<'_, crate::SharedState>,
) -> Result<GenericResponse, String> {
    Ok(GenericResponse {
        success: false,
        error: Some("Query profiling requires QoreDB Pro".into()),
    })
}

#[cfg(feature = "pro")]
#[tauri::command]
pub async fn export_profiling(
    state: State<'_, crate::SharedState>,
) -> Result<ExportResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };
    let data = interceptor.export_profiling();
    Ok(ExportResponse {
        success: true,
        data: Some(data),
        error: None,
    })
}

#[cfg(not(feature = "pro"))]
#[tauri::command]
pub async fn export_profiling(
    _state: State<'_, crate::SharedState>,
) -> Result<ExportResponse, String> {
    Ok(ExportResponse {
        success: false,
        data: None,
        error: Some("Query profiling requires QoreDB Pro".into()),
    })
}

/// Gets all safety rules (built-in + custom)
#[tauri::command]
pub async fn get_safety_rules(
    state: State<'_, crate::SharedState>,
) -> Result<SafetyRulesResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };

    let rules = interceptor.get_safety_rules();

    Ok(SafetyRulesResponse {
        success: true,
        rules,
        error: None,
    })
}

/// Adds a custom safety rule (Pro only)
#[cfg(feature = "pro")]
#[tauri::command]
pub async fn add_safety_rule(
    state: State<'_, crate::SharedState>,
    rule: SafetyRule,
) -> Result<SafetyRulesResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };

    match interceptor.add_safety_rule(rule) {
        Ok(()) => {
            let rules = interceptor.get_safety_rules();
            Ok(SafetyRulesResponse {
                success: true,
                rules,
                error: None,
            })
        }
        Err(e) => Ok(SafetyRulesResponse {
            success: false,
            rules: vec![],
            error: Some(e),
        }),
    }
}

#[cfg(not(feature = "pro"))]
#[tauri::command]
pub async fn add_safety_rule(
    _state: State<'_, crate::SharedState>,
    _rule: SafetyRule,
) -> Result<SafetyRulesResponse, String> {
    Ok(SafetyRulesResponse {
        success: false,
        rules: vec![],
        error: Some("Custom safety rules require QoreDB Pro".into()),
    })
}

/// Updates an existing safety rule (Pro only)
#[cfg(feature = "pro")]
#[tauri::command]
pub async fn update_safety_rule(
    state: State<'_, crate::SharedState>,
    rule: SafetyRule,
) -> Result<SafetyRulesResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };

    match interceptor.update_safety_rule(rule) {
        Ok(()) => {
            let rules = interceptor.get_safety_rules();
            Ok(SafetyRulesResponse {
                success: true,
                rules,
                error: None,
            })
        }
        Err(e) => Ok(SafetyRulesResponse {
            success: false,
            rules: vec![],
            error: Some(e),
        }),
    }
}

#[cfg(not(feature = "pro"))]
#[tauri::command]
pub async fn update_safety_rule(
    _state: State<'_, crate::SharedState>,
    _rule: SafetyRule,
) -> Result<SafetyRulesResponse, String> {
    Ok(SafetyRulesResponse {
        success: false,
        rules: vec![],
        error: Some("Custom safety rules require QoreDB Pro".into()),
    })
}

/// Removes a custom safety rule (Pro only)
#[cfg(feature = "pro")]
#[tauri::command]
pub async fn remove_safety_rule(
    state: State<'_, crate::SharedState>,
    rule_id: String,
) -> Result<SafetyRulesResponse, String> {
    let interceptor = {
        let state = state.lock().await;
        Arc::clone(&state.interceptor)
    };

    match interceptor.remove_safety_rule(&rule_id) {
        Ok(()) => {
            let rules = interceptor.get_safety_rules();
            Ok(SafetyRulesResponse {
                success: true,
                rules,
                error: None,
            })
        }
        Err(e) => Ok(SafetyRulesResponse {
            success: false,
            rules: vec![],
            error: Some(e),
        }),
    }
}

#[cfg(not(feature = "pro"))]
#[tauri::command]
pub async fn remove_safety_rule(
    _state: State<'_, crate::SharedState>,
    _rule_id: String,
) -> Result<SafetyRulesResponse, String> {
    Ok(SafetyRulesResponse {
        success: false,
        rules: vec![],
        error: Some("Custom safety rules require QoreDB Pro".into()),
    })
}
